cirq/sim/density_matrix.py

```diff
@@ -71,3 +72,180 @@ def to_valid_density_matrix(
+            dtype of `out` will match that of matrix if `out` is None, otherwise
+            it will match the shape and dtype of `out`.
+
+    Returns:
+        A tuple of a list and an numpy array. The list is an array of booleans
+        corresponding to the measurement values (ordered by the indices). The
+        numpy array is the post measurement matrix. This matrix has the same
+        shape and dtype as the input matrix.
+
+    Raises:
+        ValueError if the dimension of the matrix is not compatible with a
+            matrix of n qubits.
+        IndexError if the indices are out of range for the number of qubits
+            corresponding to the density matrix.
+    """
+    num_qubits = _validate_num_qubits(matrix)
+    _validate_indices(num_qubits, indices)
+
+    if len(indices) == 0:
+        return ([], np.copy(matrix))
+
+    # Cache initial shape.
+    initial_shape = matrix.shape
+
+    # Calculate the measurement probabilities and then make the measurement.
+    probs = _probs(matrix, indices, num_qubits)
+    result = np.random.choice(len(probs), p=probs)
+    measurement_bits = [(1 & (result >> i)) for i in range(len(indices))]
+
+    # Calculate the slice for the measurement result.
+    result_slice = linalg.slice_for_qubits_equal_to(indices, result, num_qubits)
+    # Create a mask which is False for only the slice.
+    mask = np.ones([2] * 2 * num_qubits, dtype=bool)
+    # Remove ellipses from last element of
+    mask[result_slice * 2] = False
+
+    if out is None:
+        out = np.copy(matrix)
+    elif out is not matrix:
+        np.copyto(out, matrix)
+    # Final else: if out is matrix then matrix will be modified in place.
+
+    # Potentially reshape to tensor, and then set masked values to 0.
+    out.shape = [2] * num_qubits * 2
+    out[mask] = 0
+
+    # Restore original shape (if necessary) and renormalize.
+    out.shape = initial_shape
+    out /= probs[result]
+
+    return measurement_bits, out
+
+
+def _probs(matrix: np.ndarray, indices: List[int],
+    num_qubits: int) -> List[float]:
+    """Returns the probabilities for a measurement on the given indices."""
+    # Only diagonal elements matter.
+    all_probs = np.diagonal(
+        np.reshape(matrix, (2 ** num_qubits, 2 ** num_qubits)))
+    # Shape into a tensor
+    tensor = np.reshape(all_probs, [2] * num_qubits)
+
+    # Calculate the probabilities for measuring the particular results.
+    probs = [
+        np.sum(np.abs(tensor[linalg.slice_for_qubits_equal_to(indices, b)]))
+        for b in range(2 ** len(indices))]
+
+    # To deal with rounding issues, ensure that the probabilities sum to 1.
+    probs /= sum(probs) # type: ignore
+    return probs
+
+
+def _validate_num_qubits(matrix: np.ndarray) -> int:
+    """Validates that matrix's shape is a valid shape for qubits.
+    """
+    shape = matrix.shape
+    half_index = len(shape) // 2
+    row_size = np.prod(shape[:half_index]) if shape != (0,) else 0
```

Comment:
Why would the shape ever be `(0,)`? It should be `()` instead in the no-qubit case.